### ISSUE TITLE
Service races in shadow mode

### ISSUE THREAD

**Issue by calebcordry** [ampproject/amphtml#22414]
After moving amp analytics' variableService to be a doc level service we noticed some previously unknown (to our team) behavior about how doc level services behave in shadow mode with multiple documents.

If a service is registered using the `AMP.registerServiceForDoc()` method it seems you can not rely on the service to be available at/before an element's `buildCallback()`. I am not sure if it is always available in later lifecycle callbacks.

My specific problem was solved by switching from the sync `getServiceForDoc` to the async `getServicePromiseForDoc` 

We are a bit concerned that this problem may exist uncaught in other places. As @zhouyx said in the previous PR 

>What about other components? Do you think this is a race condition that should be fixed by the shadow doc code instead? It would be great if you could point me to where services are registered for individual doc. ...
>  My concern is that I expect it's common to encounter this race condition and there could be existing bugs with other components already.

Follow up to #22375

**Comment by calebcordry** [comment]
cc/ @zhouyx @lannka @torch2424 @jridgewell @choumx

**Comment by calebcordry** [comment]
You can test this behavior with the files in this draft PR https://github.com/ampproject/amphtml/pull/22394 The behavior I am referencing happens when you navigate from the initial page, to all subsequent pages loaded by the shell.

**Comment by jridgewell** [comment]
@choumx, can you see any error with this? I would expect these services to be ready at `buildCallback`.

**Comment by choumx** [comment]
Big thanks for the easy-to-repro test page. So here's the problem --

Closing the shadow doc disposes all services, including extension services:
https://github.com/ampproject/amphtml/blob/f95c48e0e5ad09aa40b98e15b0c7a3f1eb2f1867/src/runtime.js#L876-L886

When the second doc is attached via `AMP.attachShadowDoc()`, the following happens:
https://github.com/ampproject/amphtml/blob/f95c48e0e5ad09aa40b98e15b0c7a3f1eb2f1867/src/runtime.js#L620-L629

`installExtensionsInDoc()` operates async but `importShadowBody()` is not promise chained to it.

`importShadowBody()` causes the `<amp-analytics>` element to be built synchronously (!) upon connecting to the DOM. 

This is because the `AmpAnalytics` class "implementation" is already available after navigating to page two. It's cached on `window.ampExtendedElements`: https://github.com/ampproject/amphtml/blob/f95c48e0e5ad09aa40b98e15b0c7a3f1eb2f1867/src/custom-element.js#L254-L261

Here's the trace:
```
buildCallback (amp-analytics-0.1.js:521)
(anonymous) (amp-shadow.js:6370)
build (amp-shadow.js:6366)
build (amp-shadow.js:33753)
buildResourceUnsafe_ (amp-shadow.js:35263)
buildReadyResourcesUnsafe_ (amp-shadow.js:35243)
buildReadyResources_ (amp-shadow.js:35217)
buildOrScheduleBuildForResource_ (amp-shadow.js:35192)
upgraded (amp-shadow.js:35352)
completeUpgrade_ (amp-shadow.js:6286)
tryUpgrade_ (amp-shadow.js:6795)
connectedCallback (amp-shadow.js:6734)
importShadowBody (amp-shadow.js:47634)
(anonymous) (amp-shadow.js:20571)
attachShadowDoc_ (amp-shadow.js:20532)
attachShadowDoc (amp-shadow.js:20563)
xhr.onreadystatechange (womp-pwa.html:55)
XMLHttpRequest.send (async)
(anonymous) (VM465:1)
loadInPwa (womp-pwa.html:63)
clickHandler (womp-pwa.html:68)
```

So a naive fix is to chain `importShadowBody()` to `installExtensionsInDoc()`. This is not good since it will delay showing the shadow document until all extensions are downloaded and installed.

A more precise fix that preserves performance would be to synchronously install extensions (and their services) that we know are already available. This should be doable at some cost of code complexity.

IMO the simplest and least brittle solution is to disallow synchronous access to all extension services (even for code within the extension itself!). @calebcordry Did you see how hairy this would be?

We'll also audit to see where else this is happening.

**Comment by calebcordry** [comment]
@choumx thanks for the investigation and the detailed write up. 
I did a quick search for `getServiceForDoc` and it used quite a bit, but it seems like there are different patterns around its usage. For instance I think something like https://github.com/ampproject/amphtml/blob/b1afbe2d15ca011fe662ff39a8bbfac7fd03bebb/extensions/amp-web-push/0.1/amp-web-push-config.js#L153 might be unsafe, but something like https://github.com/ampproject/amphtml/blob/b1afbe2d15ca011fe662ff39a8bbfac7fd03bebb/extensions/amp-live-list/0.1/live-list-manager.js#L358-L361 I think is ok since they are always installing the service before?

I am not clear on the scope, but my gut feeling is that that the code using the sync version could take a big refactor if they are calling it often but must change to wait for promise resolution.

**Comment by choumx** [comment]
This is mostly fixed with #22468 and #22511. Migrating remaining sync services tracked in #22545.

**Comment by choumx** [comment]
Reopening because I think banning sync service getters is proving to be a pain and confusing (#27260, #27990). Also, this may be solved when `buildCallback` is deferred to not happen immediately after element upgrade. 

/cc @jridgewell
